extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java

```diff
@@ -903,225 +949,223 @@ public void close() {
+    abstract static class EncodedTransfer<T> implements TransferObject<Binary[]> {
         private final ChunkSource.GetContext context;
-        private ObjectChunk<String, Values> chunk;
-        private final Binary[] buffer;
+        private ObjectChunk<T, Values> chunk;
+        private Binary[] buffer;
         private final ColumnSource<?> columnSource;
 
-        StringTransfer(ColumnSource<?> columnSource, int targetSize) {
+        /**
+         * The maximum size of data to be stored in a single page. This is not a strict limit though.
+         */
+        private final int targetPageSize;
+
+        /**
+         * Index of next object from the chunk to be buffered
+         */
+        private int currentChunkIdx;
+
+        EncodedTransfer(ColumnSource<?> columnSource, int maxValuesPerPage, int targetPageSize) {
             this.columnSource = columnSource;
-            this.buffer = new Binary[targetSize];
-            context = this.columnSource.makeGetContext(targetSize);
+            this.buffer = new Binary[maxValuesPerPage];
+            context = this.columnSource.makeGetContext(maxValuesPerPage);
+            this.targetPageSize = targetPageSize;
         }
 
         @Override
-        public void propagateChunkData() {
-            for (int i = 0; i < chunk.size(); i++) {
-                String value = chunk.get(i);
-                buffer[i] = value == null ? null : Binary.fromString(value);
+        final public boolean hasMoreDataToBuffer() {
+            return ((chunk != null) && (currentChunkIdx < chunk.size()));
+        }
+
+        @Override
+        final public int bufferAllFetchedData() {
+            if (!hasMoreDataToBuffer()) {
+                return 0;
+            }
+            while (currentChunkIdx < chunk.size()) {
+                final T value = chunk.get(currentChunkIdx);
+                buffer[currentChunkIdx++] = value == null ? null : encodeToBinary(value);
             }
+            return chunk.size();
         }
 
         @Override
-        public Binary[] getBuffer() {
-            return buffer;
+        final public int bufferFetchedData() {
+            if (!hasMoreDataToBuffer()) {
+                return 0;
+            }
+            if (currentChunkIdx != 0) {
+                // Clear any old buffered data
+                Arrays.fill(buffer, null);
+            }
+            int size = 0;
+            int bufferIdx = 0; // Stores the number of chunks buffered in this iteration
+            while (currentChunkIdx < chunk.size()) {
+                final T value = chunk.get(currentChunkIdx++);
+                if (value == null) {
+                    buffer[bufferIdx++] = null;
+                    continue;
+                }
+                Binary binaryVal = encodeToBinary(value);
+                buffer[bufferIdx++] = binaryVal;
+                size += binaryVal.length();
+                if (size > targetPageSize) {
+                    break;
+                }
```

Comment:
I agree with the >= comment.

For the optimisitic encoding, I tried that approach (it might be in one of the old commits) but then I was looking at how pyarrow does it and they are even more relaxed with the limit. They put elements in chunks of 100 in a page and then check if they have exceeded the limit. So I thought of keeping the implementation simpler by checking after buffering.
I can revert to the checking before buffering approach if you think that would be more appropriate.